Auto switching light/dark theme according to system settings

Yesterday17 opened the pull request:
Resolves #4970
Resolves #5326 

## Checklist

- [x] Added `themes` configuration
- [x] ~~Added an action to toggle between different theme modes~~ Removed as `auto` mode has been implemented
- [x] Added appearance observer to GPUI
- [x] Subscribe appearance change and change theme

Yesterday17 committed:
Add `theme_dark` and `theme_mode` settings

Yesterday17 committed:
Add action to toggle theme mode

Yesterday17 committed:
Renames

Yesterday17 commented:
@cla-bot check

Yesterday17 committed:
Add appearance observer in gpui

Yesterday17 committed:
Observe appearance change and update setting

Yesterday17 committed:
Refactor setting structure

Yesterday17 committed:
Set system appearance to global

Yesterday17 committed:
Combine `themes.auto_switch` with `themes.mode`

Yesterday17 committed:
Move common logic to `set_system_is_dark_mode`

Yesterday17 committed:
Revert default setting modify

Yesterday17 committed:
Do not expose SystemAppearance

Yesterday17 committed:
Code cleanup

Yesterday17 committed:
More code cleanups

TheDarkchip commented:
When the editor is launched in a mode that is not equal to the system appearance, it won't change themes. Only when the system appearance transition is triggered the theme is switched.

Maybe we can introduce a callback on the launch of the editor that checks for the current system appearance?

mikayla-maki commented on crates/gpui/src/window.rs:
```diff
@@ -387,6 +390,15 @@ impl Window {
             })
         });
 
+        platform_window.on_appearance_changed({
```
Oh wow, I hadn't realized we had already wired up the platform callback! :D

mikayla-maki committed:
Adjust setting to be strongly typed, for better completions
Avoid pinging the file system when switching themes on appearance
Fix a bug where the theme would not be set correctly initially

mikayla-maki committed:
Merge branch 'main' into add-appearance-switcher

mikayla-maki reviewed (dismissed):
This PR is awesome! I did a few tiny touch ups and adjusted how the theme is reloaded. As soon as CI passes, I'll merge this. Great work :)

mikayla-maki committed:
Fix tests

maxdeviant commented:
> This PR is awesome! I did a few tiny touch ups and adjusted how the theme is reloaded. As soon as CI passes, I'll merge this. Great work :)

@mikayla-maki Let's hold off on merging this one until I figure out how it works into the theme stuff I'm working on now.

Want to make sure we have a cohesive narrative around theming.

maxdeviant commented on crates/theme/src/settings.rs:
```diff
@@ -28,11 +28,68 @@ pub struct ThemeSettings {
+}
+
+impl ThemeSettings {
+    pub fn reload_theme(system_is_dark: bool, cx: &mut AppContext) {
```
Do we need to pass in `system_is_dark` if we can read it from the context?

It looks like at the call sites we always just pass in `cx.is_dark_mode()`.

maxdeviant commented on crates/theme/src/settings.rs:
```diff
@@ -52,13 +109,67 @@ pub struct ThemeSettingsContent {
+    /// Set the themes to use for light and dark mode. Note that this
+    /// Supersedes the `theme` setting.
+    #[serde(default)]
+    pub themes: Option<ThemeFeatures>,
```
I think it would be nice if we could have these settings live under the same `theme` key, since they're mutually-exclusive with it.

This should be possible by introducing a new enum to use as the value for `theme`:

```
#[serde(untagged)]
enum ThemeNameOrDynamic {
    Theme(String),
    Dynamic(ThemeFeatures)
}
```

maxdeviant commented on crates/theme/src/settings.rs:
```diff
@@ -52,13 +109,67 @@ pub struct ThemeSettingsContent {
+}
+
+impl ThemeFeatures {
+    pub fn is_dark_mode(&self, system_is_dark: bool) -> bool {
```
Instead of having `is_dark_mode`, could we have an `appearance` method that returns an [`Appearance`](https://github.com/zed-industries/zed/blob/9cb5a84b8d37062bdd0f0afe7aeae64f714d1568/crates/theme/src/theme.rs#L34-L37)?

maxdeviant commented on crates/theme/src/settings.rs:
```diff
@@ -52,13 +109,67 @@ pub struct ThemeSettingsContent {
+        match self.mode {
+            Some(ThemeMode::Auto) => system_is_dark,
+            Some(ThemeMode::Dark) => true,
+            Some(ThemeMode::Light) | _ => false,
```
Any reason to not just use `None` here instead of a wildcard? It looks like that should be the only missing case:

```suggestion
            Some(ThemeMode::Light) | None => false,
```

maxdeviant commented on crates/theme/src/settings.rs:
```diff
@@ -154,6 +265,19 @@ pub fn reset_font_size(cx: &mut AppContext) {
     }
 }
 
+struct SystemAppearance(bool);
```
I think storing an `Appearance` here would be clearer than storing a `bool`:

```suggestion
struct SystemAppearance(Appearance);
```

TheDarkchip commented:
While the detection on editor startup now works, there is some trouble left with more than one system theme change.

Notably the themes.list() of the ThemeRegistry returns only one theme (One Dark).
As follows the .get(name) method also fails on a differing name, resulting in no theme change, even when the system theme changes.

maxdeviant commented:
> I think I'm going to get this shipped @Yesterday17, thank you for getting the ball rolling :D

@mikayla-maki I was thinking it'd be nice to get the GPUI changes landed first, as those seem pretty independent.

And then I'd be down to pair on how we expose this in the settings, as I think there's some refinement we could do there.

mikayla-maki commented:
@maxdeviant which GPUI changes are you referring to?

Yesterday17 committed:
Fix issues mentioned in code review

maxdeviant committed:
Merge branch 'main' into pr/6881

maxdeviant committed:
Clean up after merge

maxdeviant committed:
Fix post-merge issues

maxdeviant committed:
Add conversion from `WindowAppearance` to `Appearance`

maxdeviant committed:
Reduce boolean treatment of appearance

maxdeviant committed:
Merge branch 'main' into pr/6881

maxdeviant committed:
Remove unused method

maxdeviant committed:
Add back blank line

maxdeviant commented:
@Yesterday17 Thank you for your work on this!

We ended up taking a slightly different approach with #7404, but we appreciate your jumping in and getting the ball rolling on this feature 🙂

mikayla-maki committed:
merge in main to stop notifications
